### PR TITLE
[13.0][ADD] mrp_unbuild_cust_alu_analytics

### PR DESCRIPTION
**Pull request by adriresu**

### CODE REVIEW

**Comment by adriresu** [comment]
TODO: 
The unlink method should trigger when deleting a record from bom_quants_ids not when deleting from bom_quants_total_ids.
Currently commented out.

**Comment by dalonsod** [comment]
Related to last change, has frontend (incidence creation and editing) behavior changed?

File: mrp_unbuild_cust_alu_analytics/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# © 2024 Solvos Consultoría Informática (<http://www.solvos.es>)
+# License LGPL-3.0 (https://www.gnu.org/licenses/lgpl-3.0.html)
+{
+    "name": "Mrp Unbuild Cust Alu Analytics",
+    "summary": """
+        Add incidences and process type to Unbuild, add custom report, add analytics to unbuilds
+    """,
+    "author": "Solvos",
+    "license": "LGPL-3",
+    "version": "13.0.1.0.0",
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/__manifest__.py:10]: Use AGPL-3

File: mrp_unbuild_cust_alu_analytics/data/ir_cron.xml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<odoo>
+    <record id="cron" model="ir.cron">
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/data/ir_cron.xml]: Cron record should always be created under a `<data noupdate="1">` section

File: mrp_unbuild_cust_alu_analytics/data/qc_test.xml [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        <record model="qc.test.question" id="test_line_25">
+            <field name="test" ref="qc_test_1"/>
+            <field name="name">Al</field>
+            <field name="type">quantitative</field>
+            <field name="min_value">0</field>
+            <field name="max_value">100</field>
+            <field name="uom_id" ref="product_uom_percentage"></field>
+        </record>
+    </data>
+</odoo>
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/data/qc_test.xml]: Missing EOF empty line

File: mrp_unbuild_cust_alu_analytics/i18n/mrp_unbuild_workflow.pot [hunk 4]
```diff
@@ -0,0 +1,506 @@
+# Translation of Odoo Server.
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/i18n/mrp_unbuild_workflow.pot]: This file should be renamed

File: mrp_unbuild_cust_alu_analytics/data/qc_test.xml [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+        <!-- UOM -->
+        <record model="uom.uom" id="product_uom_percentage">
+            <field name="name">%</field>
+            <field name="category_id" ref="product_uom_categ_percentage"></field>
+            <field name="factor" eval="1.0"/>
+            <field name="rounding" eval="0.001"/>
+        </record>
+
+        <!-- TEST -->
+        <record model="qc.test" id="qc_test_1">
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/data/qc_test.xml:19]: This record is used in code, so you must prevent its deletion. Please protect it in unlink method

File: mrp_unbuild_cust_alu_analytics/models/mrp_bom.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+# © 2024 Solvos Consultoría Informática (<http://www.solvos.es>)
+# License LGPL-3.0 (https://www.gnu.org/licenses/lgpl-3.0.html)
+
+from odoo import api, fields, models
+
+class MrpBom(models.Model):
+    _inherit = "mrp.bom"
+
+    process_type_id = fields.Many2one('mrp.unbuild.process.type', string='Process type')
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/models/mrp_bom.py]: Missing EOF empty libe

Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/models/mrp_bom.py]: Unnecessary string (Odoo will name "Process Type" this field). Don't change anything, anyway, only take it in account the next case

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+class MrpUnbuild(models.Model):
+    _inherit = "mrp.unbuild"
+
+    process_type_id = fields.Many2one('mrp.unbuild.process.type', string='Process type')
+    incidence_ids = fields.One2many(
+        comodel_name="mrp.unbuild.incidence",
+        inverse_name="unbuild_id",
+        string="Incidences",
+    )
+    analytic_ids = fields.Many2many(
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: Why this field is `Many2many`? And why computed? Inspections have an `unbuild_id` field, so it this could simply be defined as a `One2many` field, what do you think?

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+    def compute_move_analytics(self):
+        for record in self:
+            analytics = self.env["qc.inspection"].search([("unbuild_id", "=", record.id)])
+            record.analytic_ids = analytics
+
+    @api.onchange('product_id')
+    def _onchange_product_id(self):
+        if self.product_id and self.process_type_id:
+            self.bom_id = self.product_id.product_tmpl_id.bom_ids.filtered(lambda r: r.process_type_id == self.process_type_id)
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py:29]: Is it granted that this line will always return a single record (or none)?

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    def _onchange_bom_id(self):
+        if self.bom_id:
+            self.product_uom_id = self.bom_id.product_uom_id.id
+        else:
+            self.product_uom_id = False
+
+    @api.constrains('bom_quants_ids')
+    def _contrains_bom_quants_ids(self):
+        total_product_qty = sum(self.bom_quants_total_ids.mapped('total_qty'))
+        self.product_qty = total_product_qty if total_product_qty > 0 else 1
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: `@api.constrains` shouldn't be used to update values, only for validation purposes.
But this is the question: why this line? Should it be linked to this? 

https://github.com/solvosci/slv-manufacture/tree/13.0/mrp_unbuild_force_exact_qty

If you test your code without that addon installed, this method will become unneeded IMO... If I'm right, please tell me, so the correct approach should be this one:
- Improve of `mrp_unbuild_force_exact_qty` that enable a bypass confirming unbuild.
- Then, adding that bypass here, in this addon.

cc @ChristianSantamaria

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+    @api.constrains("incidence_ids")
+    def _constrains_incidences_ids(self):
+        if self.incidence_ids:
+            self.shift_stop_time = sum(self.incidence_ids.mapped('duration'))
+        else:
+            self.shift_stop_time = 0
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py:45]: the same as above: `@api.constrains` is not the proper way for this. Could you try it in other way?

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+                "&",
+                ("process_type_id", "=", record.process_type_id.id),
+                ("product_tmpl_id", "=", record.product_id.product_tmpl_id.id)
+            ]
+            domain_product_id = [
+                ("process_type_ids", "=", record.process_type_id.id),
+            ]
+        return {"domain": {"bom_id": domain_bom_id, "product_id": domain_product_id}}
+
+    def read_qc_inspection_xml(self):
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: Better name (this is a protected method, only called by a cron):
```suggestion
    def _read_qc_inspection_xml(self):
```

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+        # Find the folder and create if not exists
+        path = self.env.company.xml_analytic_path
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: This only be executed for active company for cron user (usually OdooBot). Don't change anything, but add a TODO comment, in order to make this real multi-company

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+        logger = logging.getLogger(__name__)
+        if not path:
+            logger.critical(
+                "XML Analytic Path not defined"
+            )
+            raise UserError(_("XML Analytic Path not defined"))
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: Why not create error message once? More examples below

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+                                        analytic['attributes']['ElementName']: float(element['children'][0]['text'])
+                                    })
+
+                    #Process xml data
+                    if unbuild_id:
+                        if unbuild_id.state == 'done':
+                            move_id = False
+                            move_line_ids = unbuild_id.produce_line_ids.move_line_ids
+                            for move_line in move_line_ids.move_id:
+                                if "Virtual Locations" in move_line.location_dest_id.complete_name:
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: This is a bad approximation... you mean "the move that has Virtual Locations/Production as destination location", don't you? If I'm right, this is strictly better and not depending on translations, etc:
```suggestion
                                if move_line.location_dest_id.usage == "production":
```

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild_bom_quants.py [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+# License LGPL-3.0 (https://www.gnu.org/licenses/lgpl-3.0.html)
+
+from odoo import api, fields, models, _
+
+class MrpUnbuildBomQuants(models.Model):
+    _inherit = "mrp.unbuild.bom.quants"
+    
+    def save_and_new(self):
+        self.ensure_one()
+        self.unbuild_id.product_qty = sum(self.unbuild_id.bom_quants_total_ids.mapped('total_qty'))
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/models/mrp_unbuild_bom_quants.py]: Related to above comment of "totally exact quantity" issue?

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild_bom_quants.py [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+        return super(MrpUnbuildBomQuants, self).save_and_new()
+
+    def save_and_close(self):
+        self.ensure_one()
+        self.unbuild_id.product_qty = sum(self.unbuild_id.bom_quants_total_ids.mapped('total_qty'))
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/models/mrp_unbuild_bom_quants.py]: Related to above comment of "totally exact quantity" issue?

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild_bom_total.py [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+class MrpBom(models.Model):
+    _inherit = "mrp.unbuild.bom.totals"
+
+    total_qty_percentage = fields.Float(
+        widget="percent",
+        compute="_compute_total_qty_percentage",
+        store=True
+    )
+
+    @api.depends("expected_qty")
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/models/mrp_unbuild_bom_total.py]: `@api.depends` for this compute method is very strange, because `expected_qty` isn't used within method, and other data that is actually used is not added to depends clause. Is it right?

File: mrp_unbuild_cust_alu_analytics/security/ir.model.access.csv [hunk 18]
```diff
@@ -0,0 +1,3 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_mrp_unbuild_process_type,access_mrp_unbuild_process_type,model_mrp_unbuild_process_type,base.group_user,1,1,1,1
+access_mrp_unbuild_incidence,access_mrp_unbuild_incidence,model_mrp_unbuild_incidence,base.group_user,1,1,1,1
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/security/ir.model.access.csv]: Missing EOF empty line

File: mrp_unbuild_cust_alu_analytics/security/ir.model.access.csv [hunk 19]
```diff
@@ -0,0 +1,3 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_mrp_unbuild_process_type,access_mrp_unbuild_process_type,model_mrp_unbuild_process_type,base.group_user,1,1,1,1
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/security/ir.model.access.csv]: Should any user create and delete process types? cc @ChristianSantamaria

File: mrp_unbuild_cust_alu_analytics/views/mrp_bom_views.xml [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+        </record>
+
+        <!-- Tree View -->
+        <record id="mrp_bom_form_process_type_tree_view" model="ir.ui.view">
+            <field name="name">mrp.bom.view.tree (process_id)</field>
+            <field name="model">mrp.bom</field>
+            <field name="inherit_id" ref="mrp.mrp_bom_tree_view"/>
+            <field name="arch" type="xml">
+                <xpath expr="//field[@name='type']" position="before">
+                    <field name="process_type_id" optional="0"/>
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/views/mrp_bom_views.xml]: Optional, Show or Hide?

File: mrp_unbuild_cust_alu_analytics/views/mrp_bom_views.xml [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+                </xpath>
+            </field>
+        </record>
+    </data>
+</odoo>
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/views/mrp_bom_views.xml]: Missing EOF empty line

File: mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_bom_total_views.xml [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+            <field name="model">mrp.unbuild.bom.totals</field>
+            <field name="inherit_id" ref="mrp_unbuild_bom_cust_qty.mrp_unbuild_bom_total_view_tree"/>
+            <field name="arch" type="xml">
+                <xpath expr="//field[@name='expected_qty']" position="after">
+                    <field name="total_qty_percentage" widget="percentage"/>
+                </xpath>
+            </field>
+        </record>
+    </data>
+</odoo>
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_bom_total_views.xml]: Missing EOF empty line

File: mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_process_type_views.xml [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+    <record id="action_mrp_process" model="ir.actions.act_window">
+        <field name="name">MRP Process</field>
+        <field name="res_model">mrp.unbuild.process.type</field>
+        <field name="view_mode">tree,form</field>
+    </record>
+
+    <!-- Menu Item -->
+    <menuitem id="menu_mrp_process" name="Process Types" parent="mrp.menu_mrp_configuration" action="action_mrp_process"/>
+
+</odoo>
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_process_type_views.xml]: Missing EOF empty line

File: mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_views.xml [hunk 24]
```diff
@@ -0,0 +1,95 @@
...
+        </record>
+
+        <!-- Tree View -->
+        <record id="mrp_unbuild_form_process_type_tree_view" model="ir.ui.view">
+            <field name="name">mrp.unbuild.view.tree (process_id)</field>
+            <field name="model">mrp.unbuild</field>
+            <field name="inherit_id" ref="mrp.mrp_unbuild_tree_view"/>
+            <field name="arch" type="xml">
+                <xpath expr="//field[@name='product_id']" position="before">
+                    <field name="process_type_id" optional="0"/>
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_views.xml]: Show or Hide?

File: mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_views.xml [hunk 25]
```diff
@@ -0,0 +1,95 @@
...
+            </field>
+        </record>
+
+        <record id="mrp_unbuild_form_group_shift_mrp_unbuild_workflow" model="ir.ui.view">
+            <field name="name">mrp.unbuild.form (group_shift) (mrp_unbuild_workflow)</field>
+            <field name="model">mrp.unbuild</field>
+            <field name="inherit_id" ref="mrp_unbuild_advanced.mrp_unbuild_form_view_user"/>
+            <field name="arch" type="xml">
+                <xpath expr="//notebook" position="before">
+                    <group
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_views.xml]: Why these changes? If needed, we should consider making them as PR improvement in original module.... cc @ChristianSantamaria

File: mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_views.xml [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+                    </page>
+                    <page string="Analytics">
+                        <field name="analytic_ids" class="oe_editable_tree_only">
+                        </field>
+                    </page>
+                </xpath>
+            </field>
+        </record>
+    </data>
+</odoo>
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_views.xml]: Missing EOF empty line

File: mrp_unbuild_cust_alu_analytics/views/res_config_settings_view.xml [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+                                </div>
+                            </div>
+                        </div>
+                    </div>
+                </div>
+            </xpath>
+        </field>
+    </record>
+
+</odoo>
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/views/res_config_settings_view.xml]: Missing EOF empty line

File: requirements.txt [hunk 28]
```diff
@@ -0,0 +1,2 @@
+lxml==4.3.2
+openpyxl==3.1.2
```
Comment (dalonsod) [inline requirements.txt]: Missing EOF empty line

File: requirements.txt [hunk 29]
```diff
@@ -0,0 +1,2 @@
+lxml==4.3.2
```
Comment (dalonsod) [inline requirements.txt]: lxml is already an Odoo requirement (see https://github.com/odoo/odoo/blob/4cc086d330b73514fbc64a7fcf22d8a7a9f1b691/requirements.txt#L14). Why this?

**Review by dalonsod — CHANGES_REQUESTED** [review]
See comments, only initial code review

**Comment by adriresu** [comment]
> Related to last change, has frontend (incidence creation and editing) behavior changed?

Finally any change was required, and It works like before, however, all the incidences created before this commit going to continue existing in DB but inaccessible from odoo views.

**Comment by dalonsod** [comment]
That's right, don't worry, because we're already in testing step. But when we'll update in AluDev Environment please report other users, because example data for this should be re-created.

File: mrp_unbuild_cust_alu_analytics/__manifest__.py (hunk 1 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/__manifest__.py:10]: Done

File: mrp_unbuild_cust_alu_analytics/data/ir_cron.xml (hunk 2 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/data/ir_cron.xml]: Done

File: mrp_unbuild_cust_alu_analytics/data/qc_test.xml (hunk 3 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/data/qc_test.xml]: Done

File: mrp_unbuild_cust_alu_analytics/i18n/mrp_unbuild_workflow.pot (hunk 4 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/i18n/mrp_unbuild_workflow.pot]: Done, but i need to recreate translations, too many changes from last translation

File: mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_views.xml (hunk 26 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_views.xml]: Done

File: mrp_unbuild_cust_alu_analytics/views/res_config_settings_view.xml (hunk 27 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/views/res_config_settings_view.xml]: Done

File: requirements.txt (hunk 28 above)
Comment (adriresu) [reply requirements.txt]: Done

File: requirements.txt (hunk 29 above)
Comment (adriresu) [reply requirements.txt]: mb, I didn't see that it was on requirenments, otherwise it is specified to python 3.7 versions or older, locally i'm using python 3.8

Comment (dalonsod) [reply requirements.txt]: nope, Odoo points python 3.7 and newer, check it

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 7 above)
Comment (dalonsod) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: Still pending to answered/fixed

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 8 above)
Comment (dalonsod) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py:29]: Pending to be answered

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 9 above)
Comment (dalonsod) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: Pending to be answered/fixed

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 10 above)
Comment (dalonsod) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py:45]: Pending to be answered/fixed

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 11 above)
Comment (dalonsod) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: Pending to be answered/fixed

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 12 above)
Comment (dalonsod) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: Pending to be answered

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 13 above)
Comment (dalonsod) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: Pending to be answered/fixed

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 14 above)
Comment (dalonsod) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: Pending to be answered/fixed

File: requirements.txt (hunk 29 above)
Comment (adriresu) [reply requirements.txt]: Yes, that's true, sorry

File: mrp_unbuild_cust_alu_analytics/data/qc_test.xml (hunk 5 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/data/qc_test.xml:19]: Done

**Comment by adriresu** [comment]
At the last commit, datetime is converted based on user timezone, but timezone should be removed before printing in Excel.
Do u know some better practice to get this? @dalonsod

**Comment by dalonsod** [comment]
Check this:
https://github.com/odoo/odoo/blob/4cc086d330b73514fbc64a7fcf22d8a7a9f1b691/odoo/fields.py#L1812-L1837

That method is used along some Odoo addons

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 8 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py:29]: By the user's side, I think so, but I can't assure.
Should it be protected anyway?

File: mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_views.xml (hunk 25 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_views.xml]: Solved

File: mrp_unbuild_cust_alu_analytics/models/mrp_bom.py (hunk 6 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/models/mrp_bom.py]: Solved

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 11 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: Solved

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 13 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: Remmoved logger, with UserError is enought

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 14 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: Solved

File: mrp_unbuild_cust_alu_analytics/security/ir.model.access.csv (hunk 18 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/security/ir.model.access.csv]: Solved

**Comment by ChristianSantamaria** [comment]
@adriresu @dalonsod 
I have added some improvements from today's meeting to the addon, I have tested it and it works correctly but it is not a stable version since I have done it running so as not to crash with AR. This addon the following:
- New menus to view incidents
- New incident type model
- Some translations
- Delete some fields from the incidents that were not wanted
- And change the name of the field from Stage to State

There are still some things to do, I tried but that generation of xlsx is difficult to follow... I would need two things:
- Add the name of the Unbuild in the second column
- Add between "Rest Break Time" and "Gross Time" a new column for a new field (Which I haven't done yet...) that will be "Scheduled Time" I will create this field in the other addon...

If you have any questions, tell me

**Comment by adriresu** [comment]
In my previous commit, I split the display_name with string split to get only the name without the reference_code, However this solution has a limitation, which is when the name itself has brackets (I splitted the display_name based on brackets)
I considered regex to solve this, but i don't know if there is some built-in odoo solution to fix this.
@dalonsod

**Comment by dalonsod** [comment]
Have you tried simply get `name` from product, instead of `display_name`?

imagen

**Comment by adriresu** [comment]
> Have you tried simply get `name` from product, instead of `display_name`?
> 
> imagen

it could work, but for subproducts, name dont has the subproduct tags:
image

**Comment by dalonsod** [comment]
OK, I know. Please add a `TODO` to code with this, and keep current workaround. I'm sure I had the same problem in other development, but I don't remember where, so I can't add an alternative solution right now.

**Comment by dalonsod** [comment]
I've finally found it!!

imagen

**Comment by adriresu** [comment]
> I've finally found it!!
> 
> imagen

I'm going to try this

**Comment by adriresu** [comment]
Last commit done to fit with #60

**Comment by adriresu** [comment]
Reference code is not protected on xlsx exports
image

It is necessary to be protected? @ChristianSantamaria

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 10 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py:45]: I tried with api.onchange, but looks like it doesn't work with one2many tree views

File: mrp_unbuild_cust_alu_analytics/views/mrp_bom_views.xml (hunk 20 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/views/mrp_bom_views.xml]: Hide

File: mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_views.xml (hunk 24 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_views.xml]: Hide

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 7 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: Yes better as One2Many. Change done

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 9 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: Done, It wasn't necesary

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 10 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py:45]: Changed, It works locally

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 12 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: Done

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild_bom_quants.py (hunk 15 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild_bom_quants.py]: Yes, removed at last commit

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild_bom_quants.py (hunk 16 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild_bom_quants.py]: Same, removed at last commit

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild_bom_total.py (hunk 17 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild_bom_total.py]: Changed expected_qty to the 2 fields mentioned at the function

File: mrp_unbuild_cust_alu_analytics/security/ir.model.access.csv (hunk 19 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/security/ir.model.access.csv]: They can´t, limited process type and incidence type to mrp groups

File: mrp_unbuild_cust_alu_analytics/views/mrp_bom_views.xml (hunk 21 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/views/mrp_bom_views.xml]: Done

File: mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_bom_total_views.xml (hunk 22 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_bom_total_views.xml]: Done

File: mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_process_type_views.xml (hunk 23 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_process_type_views.xml]: Done

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py [hunk 30]
```diff
@@ -0,0 +1,151 @@
...
+    process_type_id = fields.Many2one('mrp.unbuild.process.type', string='Process type')
+    incidence_ids = fields.One2many(
+        comodel_name="mrp.unbuild.incidence",
+        inverse_name="unbuild_id",
+        string="Incidences",
+    )
+    analytic_ids = fields.One2many(
+        comodel_name="qc.inspection",
+        string="Analytics",
+        store=False
```
Comment (dalonsod) [inline mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: `inverse_name` is missing (maybe is still linking data, but I don't really how it works). `store=False` does not make sense IMO. And you should add `copy=False`, in order to prevent copying analytics when duplicating an unbulld

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 8 above)
Comment (dalonsod) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py:29]: If you see `super()` implementation:

https://github.com/odoo/odoo/blob/4cc086d330b73514fbc64a7fcf22d8a7a9f1b691/addons/mrp/models/mrp_unbuild.py#L106

it's already non-multi, so your code is coherent and there's no need to fix anything :+1:

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 13 above)
Comment (dalonsod) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: Not agree. This code will be fired within a cron execution, not a user action. Who will catch those error and notice anyone. The best way should be wrapping it with try-except statement and logging errors.

@ChristianSantamaria what do you think?

File: mrp_unbuild_cust_alu_analytics/views/mrp_bom_views.xml (hunk 20 above)
Comment (dalonsod) [reply mrp_unbuild_cust_alu_analytics/views/mrp_bom_views.xml]: OK, then let's use `hide` instead of `0`.

File: mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_views.xml (hunk 24 above)
Comment (dalonsod) [reply mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_views.xml]: OK, then let's use `hide` instead of `0`.

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 30 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: From what I have read, if any inverse_name is provided; it will link anyway but  the Many2one inverse field will not be created.
But yes, the correct is as u said, mostly to be bidirectional.

Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: Fixed

File: mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py (hunk 13 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/models/mrp_unbuild.py]: I made a general try-except clause with nested try-except fired like this:
image

File: mrp_unbuild_cust_alu_analytics/views/mrp_bom_views.xml (hunk 20 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/views/mrp_bom_views.xml]: Done

File: mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_views.xml (hunk 24 above)
Comment (adriresu) [reply mrp_unbuild_cust_alu_analytics/views/mrp_unbuild_views.xml]: Done

**Review by dalonsod — APPROVED** [review]
At last!!
